clients/application/src/main/java/org/eclipse/hono/application/client/AsyncCommandClient.java

```diff
@@ -0,0 +1,133 @@
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.hono.application.client;
+
+import java.util.Map;
+
+import org.eclipse.hono.client.ServiceInvocationException;
+
+import io.opentracing.SpanContext;
+import io.vertx.core.Future;
+import io.vertx.core.buffer.Buffer;
+
+/**
+ * A client for sending asynchronous commands.
+ *
+ * @see <a href="https://www.eclipse.org/hono/docs/api/command-and-control/">
+ *      Command &amp; Control API for AMQP 1.0 Specification</a>
+ * @see <a href="https://www.eclipse.org/hono/docs/api/command-and-control-kafka/">
+ *      Command &amp; Control API for Kafka Specification</a>
+ */
+public interface AsyncCommandClient {
+    /**
+     * Sends an async command to a device, i.e. there is no immediate response expected from the device, but
+     * asynchronously via a separate consumer.
+     * <p>
+     * A device needs to be (successfully) registered before a client can upload any data for it. The device also needs
+     * to be connected for a successful delivery.
+     *
+     * @param deviceId The device to send the command to.
+     * @param command The command name.
+     * @param data The command data to send to the device or {@code null} if the command has no input data.
+     * @param correlationId The identifier to use for correlating the response with the request. Note: This parameter is
+     *            security sensitive. To ensure secure request response mapping choose correlationId carefully, e.g.
+     *            {@link java.util.UUID#randomUUID()}.
+     * @param replyId This identifier is to be set depending upon the underlying implementation (i.e AMQP or Kafka based).
+     *                <ul>
+     *                <li>In Kafka based implementation, this is not used and hence it can be set as {@code null}.</li>
+     *                <li>In AMQP based implementation, this is an arbitrary string which gets used for the response 
+     *                link address in the form of <em>command_response/${tenantId}/${replyId}</em>. Must match the
+     *                {@code replyId} passed to the command response receiver. The replyId must not be {@code null}</li>
+     *                </ul>
+     * @return A future indicating the result of the operation.
+     *         <p>
+     *         If the command was accepted, the future will succeed.
+     *         <p>
+     *         The future will fail with a {@link ServiceInvocationException} if the command could not
+     *         be forwarded upstream.
+     * @throws NullPointerException if deviceId, command or correlationId is {@code null}. 
+     *                              Also if the replyId is {@code null} provided the implementation is AMQP based.
+     */
+    Future<Void> sendAsyncCommand(String deviceId, String command, Buffer data, String correlationId, String replyId);
```

Comment:
FMPOV only the last method is necessary. I don't see enough benefit in sparing the `properties` and `context` parameters to justify the increased complexity that comes with overloading.

Reply:
I tried to keep the methods as close to the old `AsyncCommandClient` considering the users who might be already using those methods.

Reply:
A good point. On the other hand, this is our chance to improve the user-facing interface as users need to adapt client code anyways.